### PR TITLE
feat: support setting headers from file

### PR DESCRIPTION
**Pull request by Noroth**
## Summary by CodeRabbit

* **New Features**
  * Request headers can be sourced from files with configurable refresh intervals; values are cached and refreshed at runtime.

* **Behavioral Changes**
  * Router and header initialization now accept an explicit context and logger for lifecycle and cancellation.
  * File-backed sources are not allowed for response headers; missing file entries record errors and do not set headers.

* **Documentation**
  * Configuration schema and fixtures updated to validate file-backed header sources and refresh-interval semantics.

* **Tests**
  * Expanded tests cover loading, refresh behavior, precedence, caching, and edge cases.

## Checklist

- [ ] I have discussed my proposed changes in an issue and have received approval to proceed.
- [ ] I have followed the coding standards of the project.
- [ ] Tests or benchmarks have been added or updated.
- [ ] Documentation has been updated on [https://github.com/wundergraph/docs-website](https://github.com/wundergraph/docs-website).
- [ ] I have read the [Contributors Guide](https://github.com/wundergraph/cosmo/blob/main/CONTRIBUTING.md).

## Open Source AI Manifesto

This project follows the principles of the [Open Source AI Manifesto](https://human-oss.dev). Please ensure your contribution aligns with its principles.

### CODE REVIEW

File: router/core/header_rule_engine.go [hunk 1]
```diff
@@ -341,6 +358,107 @@ func (h *HeaderPropagation) compileExpressionRules(requestRules []*config.Reques
 	return nil
 }
 
+func (h *HeaderPropagation) setupFileSourceRules(logger *zap.Logger, requestRules []*config.RequestHeaderRule) error {
+	for _, rule := range requestRules {
+		if !isValidFileSourceRule(rule) {
+			continue
```
Comment (SkArchon) [inline router/core/header_rule_engine.go:396]: probably good to add a warning, since this means the user configuration is off?

File: router/core/header_rule_engine.go [hunk 2]
```diff
@@ -341,6 +358,107 @@ func (h *HeaderPropagation) compileExpressionRules(requestRules []*config.Reques
...
+		}
+
+		if rule.FromFile.RefreshInterval <= 0 {
+			rule.FromFile.RefreshInterval = 30 * time.Second
+		}
+
+		_, err := os.Stat(rule.FromFile.Path)
+		if err != nil {
+			if os.IsNotExist(err) {
```
Comment (SkArchon) [inline router/core/header_rule_engine.go]:
```// This function predates [errors.Is]. It only supports errors returned by
// the os package. New code should use errors.Is(err, fs.ErrNotExist).```

From the docs

File: router/core/header_rule_engine.go [hunk 3]
```diff
@@ -341,6 +358,107 @@ func (h *HeaderPropagation) compileExpressionRules(requestRules []*config.Reques
...
+				return fmt.Errorf("file %s does not exist", rule.FromFile.Path)
+			}
+			return fmt.Errorf("error statting file %s: %w", rule.FromFile.Path, err)
```
Comment (SkArchon) [inline router/core/header_rule_engine.go]: nit: `statting` sounds a bit weird

File: router/core/header_rule_engine.go [hunk 4]
```diff
@@ -341,6 +358,107 @@ func (h *HeaderPropagation) compileExpressionRules(requestRules []*config.Reques
...
+			return fmt.Errorf("error creating watcher for file %s: %w", rule.FromFile.Path, err)
+		}
+
+		content, found := h.fileSourceContents[rule.FromFile.Path]
+		if !found {
+			return fmt.Errorf("file source content not found for file %s", rule.FromFile.Path)
+		}
+
+		content.watcher = watcher
+		go content.watcher(context.Background())
```
Comment (SkArchon) [inline router/core/header_rule_engine.go]: You will need to pass down the cancelable context from the router, as this probably does not cause the routine to stop on router reloads.

**Review by SkArchon — COMMENTED** [review]
structure ok, few comments though

File: router/core/header_rule_engine.go [hunk 5]
```diff
@@ -590,6 +708,18 @@ func (h *HeaderPropagation) applyRequestRuleToHeader(ctx *requestContext, header
...
+			content, ok := h.fileSourceContents[rule.FromFile.Path]
+			if !ok {
+				ctx.SetError(fmt.Errorf("file source content not found for file %s", rule.FromFile.Path))
+			}
+
+			content.m.Lock()
+			header.Set(rule.Name, content.buffer.String())
+			content.m.Unlock()
+			return
+		}
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/core/header_rule_engine.go:731]: @Noroth seems legit, can you check

File: router/core/header_rule_engine.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Noroth) [reply router/core/header_rule_engine.go]: fixed. Passed the cancellable router context to the function

File: router/pkg/config/config.schema.json [hunk 6]
```diff
@@ -4633,6 +4633,82 @@
...
+              "description": "The field name of the context from which to extract the value. The value is only extracted when a context is available otherwise the default value is used.",
+              "enum": ["operation_name"]
+            }
+          }
+        },
+        "from_file": {
+          "type": "object",
+          "description": "Source the header value from a file's contents. The file is loaded into memory and refreshed at refresh_interval, so request-time reads do not touch disk. Currently only valid in requests.",
+          "additionalProperties": false,
+          "required": ["path", "refresh_interval"],
```
[the remark this answers is not in the document]
Comment (Noroth) [reply router/pkg/config/config.schema.json:4653]: removed the obsolete envDefault from config.go. This is actually not working therefore I marked the field required to ensure that we have to provide a proper value

File: router/core/header_rule_engine.go [hunk 7]
```diff
@@ -158,6 +164,43 @@ type HeaderPropagation struct {
...
+	f.m.Lock()
+	defer f.m.Unlock()
+
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return fmt.Errorf("error reading file %s: %w", path, err)
+	}
+
+	f.buffer.Reset()
+	// buffer Write is guaranteed to never return an error
```
Comment (SkArchon) [inline router/core/header_rule_engine.go]: Since we are ignoring the second arg which is err, would be good to also mention "why" it will never return an error in the comment.

**Review by SkArchon — APPROVED** [review]
lgtm
